File: ts/src/receiver.ts
import type { Logger } from 'webex-message-handler';
import type { ReceiverConfig } from './config.js';
import { decompress as codecDecompress, decrypt as codecDecrypt } from './codec.js';
import { Allowlist } from './allowlist.js';
import { getLogger } from './logging.js';
import { parseResponse, parseFlags, isPause, isResume, stripBotMention } from './protocol.js';
import { getMessage, getAttachmentAction, extractCards } from './webex.js';
import { createListener, type IncomingMessage, type PlatformListener } from './listener.js';

export type MessageHandler = (slug: string, payload: string, cards: unknown[], fromSlug: string) => void | Promise<void>;
export type ControlHandler = (cmd: string) => void | Promise<void>;

export class WgrokReceiver {
  private config: ReceiverConfig;
  private allowlist: Allowlist;
  private messageHandler: MessageHandler;
  private controlHandler?: ControlHandler;
  private logger: Logger;
  private listener?: PlatformListener;
  private abortController?: AbortController;
  private chunkBuffer: Map<string, Map<number, string>> = new Map();

  constructor(config: ReceiverConfig, handler: MessageHandler, controlHandler?: ControlHandler) {
    this.config = config;
    this.allowlist = new Allowlist(config.domains);
    this.messageHandler = handler;
    this.controlHandler = controlHandler;
    this.logger = getLogger(config.debug, 'wgrok.receiver');
  }

  async listen(): Promise<void> {
    this.abortController = new AbortController();
    this.listener = createListener(this.config.platform, this.config.webexToken, this.logger);

    this.listener.onMessage(async (msg: IncomingMessage) => {
      await this.onMessage(msg).catch((err) => {
        this.logger.error(`Error handling message: ${err}`);
      });
    });

    this.logger.info(`Receiver listening for slug: ${this.config.slug}`);
    await this.listener.connect();
    this.logger.info('Receiver connected');

    await new Promise<void>((resolve) => {
      this.abortController!.signal.addEventListener('abort', () => resolve());
    });
  }

  async stop(): Promise<void> {
    this.abortController?.abort();
    if (this.listener) {
      await this.listener.disconnect();
      this.listener = undefined;
    }
    this.logger.info('Receiver stopped');
  }

  async fetchAction(actionId: string): Promise<Record<string, unknown>> {
    return getAttachmentAction(this.config.webexToken, actionId);
  }

  /** Exposed for testing with injected cards */
  async onMessageWithCards(msg: IncomingMessage, cards: unknown[]): Promise<void> {
    const sender = msg.sender;
    const text = stripBotMention(msg.text, msg.html);

    if (!this.allowlist.isAllowed(sender)) {
      this.logger.warn(`Rejected message from ${sender}: not in allowlist`);
      return;
    }

    // Handle control commands before parsing as response
    if (isPause(text)) {
      this.logger.info(`Received pause command from ${sender}`);
      if (this.controlHandler) {
        await this.controlHandler('pause');
      }
      return;
    }

    if (isResume(text)) {
      this.logger.info(`Received resume command from ${sender}`);
      if (this.controlHandler) {
        await this.controlHandler('resume');
      }
      return;
    }

    let to: string, from: string, flags: string, payload: string;
    try {
      ({ to, from, flags, payload } = parseResponse(text));
    } catch {
      this.logger.debug(`Ignoring unparseable message from ${sender}`);
      return;
    }

    if (to !== this.config.slug) {
      this.logger.debug(`Ignoring message with to "${to}" (expected "${this.config.slug}")`);
      return;
    }

    let compressed = false;
    let encrypted = false;
    let chunkSeq: number | null = null;
    let chunkTotal: number | null = null;

    try {
      ({ compressed, encrypted, chunkSeq, chunkTotal } = parseFlags(flags));
    } catch {
      this.logger.debug(`Ignoring message with invalid flags "${flags}"`);
      return;
    }

    // Handle chunked payload
    if (chunkSeq !== null && chunkTotal !== null) {
      if (chunkTotal > 999 || chunkSeq > chunkTotal || chunkSeq < 1) {
        this.logger.warn(`Invalid chunk ${chunkSeq}/${chunkTotal} from ${sender}`);
        return;
      }
      const key = `${sender}:${to}`;
      if (!this.chunkBuffer.has(key)) {
        this.chunkBuffer.set(key, new Map());
      }
      this.chunkBuffer.get(key)!.set(chunkSeq, payload);
      if (this.chunkBuffer.get(key)!.size < chunkTotal) {
        this.logger.debug(`Buffered chunk ${chunkSeq}/${chunkTotal} for to "${to}" from ${sender}`);
        return;
      }
      // All chunks received — reassemble
      const parts: string[] = [];
      for (let i = 1; i <= chunkTotal; i++) {
        parts.push(this.chunkBuffer.get(key)!.get(i)!);
      }
      payload = parts.join('');
      this.chunkBuffer.delete(key);
      this.logger.debug(`Reassembled ${chunkTotal} chunks for to "${to}" from ${sender}`);
    }

    // Decrypt if encrypted
    if (encrypted) {
      if (!this.config.encryptKey) {
        this.logger.warn(`Received encrypted message but no WGROK_ENCRYPT_KEY is set`);
        return;
      }
      try {
        payload = codecDecrypt(payload, this.config.encryptKey);
      } catch (err) {
        this.logger.warn(`Failed to decrypt message: ${err}`);
        return;
      }
    }

    // Decompress if compressed
    if (compressed) {
      try {
        payload = codecDecompress(payload);
      } catch (err) {
        this.logger.warn(`Failed to decompress message: ${err}`);
        return;
      }
    }

    if (cards.length > 0) {
      this.logger.info(`Received payload for to "${to}" from ${sender} (with ${cards.length} card(s))`);
    } else {
      this.logger.info(`Received payload for to "${to}" from ${sender}`);
    }
    await this.messageHandler(to, payload, cards, from);
  }

  private async onMessage(msg: IncomingMessage): Promise<void> {
    // For webex, fetch additional card data. For other platforms, cards come with the message
    let cards = msg.cards;
    if (msg.platform === 'webex' && msg.msgId) {
      cards = await this.fetchCards(msg.msgId);
    }
    await this.onMessageWithCards(msg, cards);
  }

  private async fetchCards(messageId: string): Promise<unknown[]> {
    if (!messageId) return [];
    try {
      const fullMsg = await getMessage(this.config.webexToken, messageId);
      return extractCards(fullMsg);
    } catch (err) {
      this.logger.debug(`Could not fetch message attachments: ${err}`);
      return [];
    }
  }
}

File: rust/src/receiver.rs
use std::collections::HashMap;
use std::sync::Mutex;
use std::time::Instant;

use reqwest::Client;
use serde_json::Value;
use tokio::sync::watch;
use webex_message_handler::{Config, DecryptedMessage, HandlerEvent, WebexMessageHandler};

use crate::allowlist::Allowlist;
use crate::codec;
use crate::config::ReceiverConfig;
use crate::logging::{get_logger, WgrokLogger};
use crate::protocol::{parse_flags, parse_response, is_pause, is_resume, strip_bot_mention};
use crate::webex;

#[derive(Debug, Clone)]
pub struct MessageContext {
    pub msg_id: String,
    pub sender: String,
    pub platform: String,
    pub room_id: String,
    pub room_type: String,
}

pub type MessageHandler = Box<dyn Fn(&str, &str, &[Value], &str, &MessageContext) + Send + Sync>;
pub type ControlHandler = Box<dyn Fn(&str) + Send + Sync>;

type ChunkKey = (String, String); // (sender, slug)

pub struct WgrokReceiver {
    config: ReceiverConfig,
    allowlist: Allowlist,
    handler: MessageHandler,
    on_control: Option<ControlHandler>,
    logger: WgrokLogger,
    client: Client,
    chunk_buffer: Mutex<HashMap<ChunkKey, HashMap<usize, String>>>,
    chunk_timestamps: Mutex<HashMap<ChunkKey, Instant>>,
}

impl WgrokReceiver {
    pub fn new(config: ReceiverConfig, handler: MessageHandler) -> Self {
        let logger = get_logger(config.debug, "wgrok.receiver");
        let allowlist = Allowlist::new(&config.domains);
        Self {
            config,
            allowlist,
            handler,
            on_control: None,
            logger,
            client: Client::new(),
            chunk_buffer: Mutex::new(HashMap::new()),
            chunk_timestamps: Mutex::new(HashMap::new()),
        }
    }

    pub fn set_control_handler(&mut self, on_control: ControlHandler) {
        self.on_control = Some(on_control);
    }

    pub async fn listen(&self, mut shutdown_rx: watch::Receiver<bool>) -> Result<(), String> {
        let ws_handler = WebexMessageHandler::new(Config {
            token: self.config.webex_token.clone(),
            ..Default::default()
        })
        .map_err(|e| format!("create handler: {}", e))?;

        let mut rx = ws_handler
            .take_event_rx()
            .await
            .ok_or("failed to take event rx")?;

        ws_handler
            .connect()
            .await
            .map_err(|e| format!("connect: {}", e))?;

        self.logger
            .info(&format!("Receiver listening for slug: {}", self.config.slug));

        loop {
            tokio::select! {
                event = rx.recv() => {
                    match event {
                        Some(HandlerEvent::MessageCreated(msg)) => {
                            let cards = self.fetch_cards(&msg.id).await;
                            self.on_message_with_cards(&msg, &cards);
                        }
                        None => break,
                        _ => {}
                    }
                }
                _ = shutdown_rx.changed() => {
                    break;
                }
            }
        }

        ws_handler.disconnect().await;
        self.logger.info("Receiver stopped");
        Ok(())
    }

    pub async fn fetch_action(&self, action_id: &str) -> Result<Value, String> {
        webex::get_attachment_action(&self.config.webex_token, action_id, &self.client).await
    }

    pub fn on_message_with_cards(&self, msg: &DecryptedMessage, cards: &[Value]) {
        let sender = &msg.person_email;
        let html = msg.html.as_deref().unwrap_or("");
        let text = strip_bot_mention(msg.text.trim(), html);

        if !self.allowlist.is_allowed(sender) {
            self.logger
                .warn(&format!("Rejected message from {}: not in allowlist", sender));
            return;
        }

        // Check for control commands before parsing as response
        if is_pause(&text) {
            self.logger.info(&format!("Received pause command from {}", sender));
            if let Some(ref handler) = self.on_control {
                handler("pause");
            }
            return;
        }

        if is_resume(&text) {
            self.logger.info(&format!("Received resume command from {}", sender));
            if let Some(ref handler) = self.on_control {
                handler("resume");
            }
            return;
        }

        let (to, from_slug, flags, mut payload) = match parse_response(&text) {
            Ok(v) => v,
            Err(_) => {
                self.logger
                    .debug(&format!("Ignoring unparseable message from {}", sender));
                return;
            }
        };

        if to != self.config.slug {
            self.logger.debug(&format!(
                "Ignoring message with slug \"{}\" (expected \"{}\")",
                to, self.config.slug
            ));
            return;
        }

        let (compressed, encrypted, chunk_seq, chunk_total) = parse_flags(&flags);

        // Handle chunking
        if let (Some(seq), Some(total)) = (chunk_seq, chunk_total) {
            if total > 100 || seq > total || seq < 1 {
                self.logger.warn(&format!("Invalid chunk {}/{} from {}", seq, total, sender));
                return;
            }
            let key = (sender.clone(), to.clone());

            // Check for chunk timeout (5 minutes)
            {
                let mut timestamps = self.chunk_timestamps.lock().unwrap();
                if let Some(ts) = timestamps.get(&key) {
                    if ts.elapsed() > std::time::Duration::from_secs(300) {
                        self.logger.warn(&format!(
                            "Discarding incomplete chunk set for {:?} (timeout after 5 minutes)", key
                        ));
                        let mut buffer = self.chunk_buffer.lock().unwrap();
                        buffer.remove(&key);
                        timestamps.remove(&key);
                        return;
                    }
                } else {
                    timestamps.insert(key.clone(), Instant::now());
                }
            }

            let mut buffer = self.chunk_buffer.lock().unwrap();
            buffer.entry(key.clone()).or_default().insert(seq, payload);
            if buffer[&key].len() < total {
                self.logger.debug(&format!(
                    "Buffered chunk {}/{} for slug \"{}\" from {}",
                    seq, total, to, sender
                ));
                return;
            }

            // Verify all indices 1..total are present before reassembly
            let all_present = (1..=total).all(|i| buffer[&key].contains_key(&i));
            if !all_present {
                self.logger.warn(&format!("Incomplete chunk set for {:?}: missing indices, discarding", key));
                buffer.remove(&key);
                drop(buffer);
                self.chunk_timestamps.lock().unwrap().remove(&key);
                return;
            }

            // All chunks received and verified — reassemble
            let chunks = buffer.remove(&key).unwrap();
            drop(buffer);
            self.chunk_timestamps.lock().unwrap().remove(&key);
            let mut assembled = String::new();
            for i in 1..=total {
                if let Some(part) = chunks.get(&i) {
                    assembled.push_str(part);
                }
            }
            self.logger.debug(&format!(
                "Reassembled {} chunks for slug \"{}\" from {}",
                total, to, sender
            ));
            payload = assembled;
        }

        // Decrypt if needed
        if encrypted {
            if let Some(key) = &self.config.encrypt_key {
                match codec::decrypt(&payload, key) {
                    Ok(plaintext) => payload = plaintext,
                    Err(e) => {
                        self.logger.warn(&format!("Decryption failed: {}", e));
                        return;
                    }
                }
            } else {
                self.logger.warn("Message is encrypted but WGROK_ENCRYPT_KEY not set, skipping");
                return;
            }
        }

        // Decompress if needed
        if compressed {
            match codec::decompress(&payload) {
                Ok(decompressed) => payload = decompressed,
                Err(e) => {
                    self.logger.warn(&format!("Decompression failed: {}", e));
                    return;
                }
            }
        }

        if !cards.is_empty() {
            self.logger.info(&format!(
                "Received payload for slug \"{}\" from {} (with {} card(s))",
                to,
                sender,
                cards.len()
            ));
        } else {
            self.logger.info(&format!(
                "Received payload for slug \"{}\" from {}",
                to, sender
            ));
        }
        let ctx = MessageContext {
            msg_id: msg.id.clone(),
            sender: sender.to_string(),
            platform: "webex".to_string(),
            room_id: msg.room_id.clone(),
            room_type: msg.room_type.clone().unwrap_or_default(),
        };
        (self.handler)(&to, &payload, cards, &from_slug, &ctx);
    }

    async fn fetch_cards(&self, message_id: &str) -> Vec<Value> {
        if message_id.is_empty() {
            return vec![];
        }
        match webex::get_message(&self.config.webex_token, message_id, &self.client).await {
            Ok(msg) => webex::extract_cards(&msg),
            Err(e) => {
                self.logger
                    .debug(&format!("Could not fetch message attachments: {}", e));
                vec![]
            }
        }
    }
}
